planner/core/plan.go

```diff
@@ -210,17 +213,36 @@ func (p *baseLogicalPlan) storeTask(prop *property.PhysicalProperty, task task)
 	p.taskMap[string(key)] = task
 }
 
-// BuildKeyInfo implements LogicalPlan BuildKeyInfo interface.
-func (p *baseLogicalPlan) BuildKeyInfo() {
-	for _, child := range p.children {
-		child.BuildKeyInfo()
-	}
-	switch p.self.(type) {
+// HasMaxOneRow returns if the LogicalPlan will output at most one row.
+func HasMaxOneRow(p LogicalPlan, childMaxOneRow []bool) bool {
+	switch p.(type) {
 	case *LogicalLock, *LogicalLimit, *LogicalSort, *LogicalSelection, *LogicalApply, *LogicalProjection:
-		p.maxOneRow = p.children[0].MaxOneRow()
+		if len(childMaxOneRow) == 1 {
+			return childMaxOneRow[0]
+		}
 	case *LogicalMaxOneRow:
-		p.maxOneRow = true
+		return true
+	case *LogicalJoin:
+		if len(childMaxOneRow) == 2 {
+			return childMaxOneRow[0] && childMaxOneRow[1]
+		}
+	}
+	return false
+}
+
+// BuildKeyInfo implements LogicalPlan BuildKeyInfo interface.
+func (p *baseLogicalPlan) BuildKeyInfo(selfSchema *expression.Schema, childSchema []*expression.Schema) {
+	childMaxOneRow := make([]bool, len(p.children))
+	for i := range p.children {
+		childMaxOneRow[i] = p.children[i].MaxOneRow()
 	}
+	p.maxOneRow = HasMaxOneRow(p.self, childMaxOneRow)
```

Comment:
For logical plans that have their own `BuildKeyInfo` but does not consider the child, this is useless? Should we put it in `buildKeyInfo`?

Reply:
Almost all of the logical plans should consider their child. If their child's maxOneRow is true, their maxOneRow should be true, even though they have their own `BuildKeyInfo`. So in every `BuildKeyInfo`, we call `p.baseLogicalPlan.BuildKeyInfo()` to check this at the very beginning.